RFS/src/main/java/com/rfs/common/S3SnapshotCreator.java

```diff
@@ -0,0 +1,152 @@
+        String targetName = "_snapshot/s3_repo";
+
+        // Assemble the request body
+        ObjectNode settings = mapper.createObjectNode();
+        settings.put("bucket", getBucketName());
+        settings.put("region", s3Region);
+        settings.put("base_path", getBasePath());
+
+        ObjectNode body = mapper.createObjectNode();
+        body.put("type", "s3");
+        body.set("settings", settings);
+
+        // Register the repo; it's fine if it already exists
+        RestClient client = new RestClient(connectionDetails);
+        String bodyString = body.toString();
+        RestClient.Response response = client.put(targetName, bodyString, false);
+        if (response.code == HttpURLConnection.HTTP_OK) {
+            logger.info("S3 Repo registration successful");
+        } else {
+            logger.error("S3 Repo registration failed");
+            throw new RepoRegistrationFailed("s3_repo");
+        }
+    }
+
+    public void createSnapshot() throws Exception {
+        // Assemble the REST path
+        String targetName = "_snapshot/s3_repo/" + getSnapshotName();
+
+        // Assemble the request body
+        ObjectNode body = mapper.createObjectNode();
+        body.put("indices", "_all");
+        body.put("ignore_unavailable", true);
+        body.put("include_global_state", true);
+
+        // Register the repo; it's fine if it already exists
+        RestClient client = new RestClient(connectionDetails);
+        String bodyString = body.toString();
+        RestClient.Response response = client.put(targetName, bodyString, false);
+        if (response.code == HttpURLConnection.HTTP_OK) {
+            logger.info("Snapshot " + getSnapshotName() + " creation successful");
+        } else {
+            logger.error("Snapshot " + getSnapshotName() + " creation failed");
+            throw new SnapshotCreationFailed(getSnapshotName());
+        }
+    }
+
+    public boolean isSnapshotFinished() throws Exception {
+        // Assemble the REST path
+        String targetName = "_snapshot/s3_repo/" + getSnapshotName();
+
+        // Check if the snapshot has finished
+        RestClient client = new RestClient(connectionDetails);
+        RestClient.Response response = client.get(targetName, false);
+        if (response.code == HttpURLConnection.HTTP_NOT_FOUND) {
+            logger.error("Snapshot " + getSnapshotName() + " does not exist");
+            throw new SnapshotDoesNotExist(getSnapshotName());
+        }
+        JsonNode responseJson = mapper.readTree(response.body);
+        JsonNode firstSnapshot = responseJson.path("snapshots").get(0);
+        JsonNode stateNode = firstSnapshot.path("state");
+        String state = stateNode.asText();
+
+        if (state.equals("SUCCESS")) {
+            return true;
+        } else if (state.equals("IN_PROGRESS")) {
+            return false;
+        } else {
+            logger.error("Snapshot " + getSnapshotName() + " has failed");
```

Comment:
We should log the failed state as well

Reply:
Ah, good catch.  I was thinking the REST Client would log that, but that might not be the case in this situation.

Reply:
Done!